src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs

```diff
@@ -132,6 +121,23 @@ private ProjectState(
         }
     }
 
+    public static ProjectState Create(
+        HostProject hostProject,
+        ProjectWorkspaceState? projectWorkspaceState = null,
+        RazorCompilerOptions compilerOptions = RazorCompilerOptions.None,
```

Comment:
> Default values scare me in general, but in particular this one seems risky. If we accidentally stop honouring feature flags in some places, it would be very likely to sneak past all of our unit and integration test gates.

Note: While I started with optional parameters, the final changes in the PR broke these out into overloads.

> I'm guessing this overload is only used by tests, perhaps it would be better to move it to a test helper instead of being here? In future I think this would be a perfect use of an static extension method defined in a test library.

You're right that the majority of these are only called by tests. In fact, the only place in product code that calls `ProjectState.Create(...)` is in `ProjectSnapshotManager` and it would be a big red flag if `ProjectState.Create(...)` were called from anywhere else. So, it's extremely unlikely that we would suddenly stop passing the provided options into `ProjectState`.

Because of the way that `ProjectState` and `DocumentState` are modeled as APIs, I really don't like moving all but one `Create(...)` overload to test helpers. Yes, they'd be a good candidate for static extension methods, but we don't have those today and the alternative isn't lovely. I'd much rather our tests call `ProjectState.Create(...)` rather than having to call `TestProjectState.Create(...)` to allow a few sensible defaults.

Reply:
To be honest, I feel like the ask is over-paranoid, especially given how bad things already are today. In main, 'ProjectState' has settable properties that could be used by product code to corrupt its state. Or, it's possible that a new `LanguageServerFeatureOptions` could be introduced that isn't immutable and changes when VS options change, which would affect the compiler flags and corrupt the `ProjectSnapshotManager`.

I didn't add these for test code convenience; I added them to allow `ProjectState` to define its own defaults when not provided more meaningful values. None of your suggestions make me feel good. Adding a `TestAccessor` is still a problem.